simulators/ethereum/pyspec/runner.go

```diff
@@ -0,0 +1,196 @@
+		"HIVE_CHAIN_ID":      "1",
+		"HIVE_SKIP_POW":      "1",
+		"HIVE_NODETYPE": 	  "full",
+	}
+	tc.updateEnv(env)
+	t0 := time.Now()
+
+	// start client (also creates an engine RPC client internally)
+	t.Log("HIVE LOG --> Starting client with Engine API.")
+	engineClient, err := engineAPI.StartClient(t, ctx, tc.genesis, env, nil)
+	if err != nil {
+		t.Fatalf("HIVE FATAL --> Can't start client with Engine API: %v", err)
+	}
+	t1 := time.Now()
+
+	// send payloads
+	for blockNumber, payload :=  range tc.payloads {
+		_, err := engineClient.NewPayloadV2(context.Background(), payload)
+		if err != nil {
+			t.Fatalf("HIVE FATAL ---> Unable to send payload[%v] in test %s: %v ", blockNumber, tc.name, err)
+		}
+	}
+	t2 := time.Now()
+
+	// update head of "beacon" chain with the latest valid response
+	latestBlockHash := *engineClient.LatestNewPayloadResponse().LatestValidHash
+	fcState := &api.ForkchoiceStateV1{HeadBlockHash: latestBlockHash}
+	engineClient.ForkchoiceUpdatedV2(ctx, fcState, nil)
+	t3 := time.Now()
+
+	// check nonce, balance & storage of accounts in final block against fixture values
+	latestBlockNumber := new(big.Int).SetUint64(engineClient.LatestNewPayloadSent().Number)
+	for account, genesisAccount :=  range tc.postAlloc {
+
+		// get nonce & balance from last block (end of test execution)
+		gotNonce, errN := engineClient.NonceAt(ctx, account, latestBlockNumber)
+		gotBalance, errB := engineClient.BalanceAt(ctx, account, latestBlockNumber)
+		if errN != nil {
+			t.Errorf("HIVE ERROR ---> Unable to call nonce from account: %v, in test %s: %v", account, tc.name, errN)
+		} else if errB != nil {
+			t.Errorf("HIVE ERROR ---> Unable to call balance from account: %v, in test %s: %v", account, tc.name, errB)
+		} 
+
+		// check final nonce & balance matches expected in fixture 
+		if genesisAccount.Nonce != gotNonce {
+			t.Errorf(`HIVE ERROR ---> Nonce recieved from account %v doesn't match expected from fixture in test %s:
+			Recieved from block: %v
+			Expected in fixture: %v`, account, tc.name, gotNonce, genesisAccount.Nonce)
+		}
+		if common.BigToHash(genesisAccount.Balance) != common.BigToHash(gotBalance) {
+			t.Errorf(`HIVE ERROR ---> Balance recieved from account %v doesn't match expected from fixture in test %s:
+			Recieved from block: %v
+			Expected in fixture: %v`, account, tc.name, gotBalance, genesisAccount.Balance)
+		}
+	
+		// check values in storage match those in fixture
+		for stPosition, stValue := range genesisAccount.Storage {
+			// get value in storage position stPosition from last block
+			gotStorage, errS := engineClient.StorageAt(ctx, account, common.BytesToHash(stPosition.Bytes()), latestBlockNumber)
+			if errS != nil {
+				t.Errorf("HIVE ERROR ---> Unable to call storage bytes from account address %v, storage position %v, in test %s: %v", account, stPosition, tc.name, errS)
+			}
+			// check recieved value against fixture value
+			if stValue != common.BytesToHash(gotStorage) {
```

Comment:
It could be worth it to convert this block into a batch request like in https://github.com/ethereum/go-ethereum/blob/ee530c0d5aa70d2c00ab5691a89ab431b73f8165/ethclient/ethclient.go#L143-L162

Reply:
I have something like this. My assumption is that by using the batch request if will speed up the overall simulation, but at the cost of extra code?

```go
// create a batch of RPC calls for each storage position
if len(genesisAccount.Storage) > 0 {
	stValues := make(map[common.Hash]common.Hash, len(genesisAccount.Storage))
	reqs := make(map[common.Hash]rpc.BatchElem, len(genesisAccount.Storage))
	// create request for every storage element
	for stPosition, _ := range genesisAccount.Storage {
		reqs[stPosition] = rpc.BatchElem{
			Method: "eth_getStorageAt",
			Args:   []interface{}{account, stPosition, "latest"},
			Result: stValues[stPosition],
		}
	}
	// batch call for each storage element
	engineClient.BatchCallContext(ctx, reqs)
	// check each request was successful and that each storage value matches expected 
	for stPosition, expValue := range genesisAccount.Storage {
		if reqs[stPosition].Error != nil {
			fmt.Errorf(`HIVE ERROR ---> Unable to call storage bytes from account address %v, storage position %v, in test %s:
				%v`, account, stPosition, tc.name, reqs[stPosition].Error)
		}
		if stValues[stPosition] != expValue {
			fmt.Errorf(`HIVE ERROR ---> Storage received from account %v doesn't match expected from fixture in test %s:
			From storage address: %v
			Received from block: %v
			Expected in fixture: %v`, account, tc.name, stPosition, stValues[stPosition], expValue)
		}
	}
}
```

Reply:
It packages everything in a single RPC call, which then the client should optimize and fetch all at the same time from its internal state DB.

So I definitely think it's worth it for bigger test cases where we check a ton of storage keys from different accounts.